# BAM to counts container


## Installation/Usage:
**Step 1**: Pull the docker image
```bash
$ docker pull biodepot/bam2deseq
```

**Step 2**: Run the docker image
```bash
$ docker run -it --rm biodepot/bam2deseq
```

**Step 3**: Run the script 

**Step A** -- BAM to counts table
```bash
    Rscript bam2counts.R [GTF File] [BAM file Directory] [Workers number]

    Arguments:
    [GTF File] a file specifying the genomic features.
    [BAM file Directory] is a directory contains BAM files.
    [Workers number] is a parameter defined number of threads, default = 1
```
**Step B** -- Counts table to DESeq
```bash
    Rscript counts2DESeq.R [Counts Table File] [Sample Table File] [Output directory]
    Arguments:
    [Counts Table File] A counts table generated by bam2counts.R 
    [Sample Table File] is a table with detailed information for each of samples that links samples to the associated FASTQ and BAM files.
    [Output directory] is the output for saving result table.
```